### PR TITLE
4760: Reset user migration status

### PR DESCRIPTION
**Pull request by kasperg**
#### Link to issue

https://platform.dandigbib.org/issues/4760

#### Description

The changes in this PR aims to provide more transparency into the user migration process when you do not have access to the underlying platform.

- Changes to the administrative settings provide functionality which can tell whether a specific user has been migrated or not.
- Improved logging can sell whether a user has been migrated or not and why.

Please check commit comments for additional details.

#### Screenshot of the result

React components | Ding2 - Mozilla Firefox 2020-02-28 17-03-51

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

### CODE REVIEW

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: a104462f7b5ae90316600bac1217ca1ffc527cfd
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: a104462f7b5ae90316600bac1217ca1ffc527cfd
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

File: modules/ding_adgangsplatformen/ding_adgangsplatformen.module [hunk 1]
```diff
@@ -434,23 +434,38 @@ function ding_adgangsplatformen_callback() {
         if (_ding_adgangsplatformen_provider_login($user_info)) {
           // Once the user is logged in then try to migrate all variations of
           // the user which might exist on the current site.
-          if (module_exists('ding_react') && is_array($user_info['attributes']['agencies'])) {
-            $user_agencies = $user_info['attributes']['agencies'];
+          if (module_exists('ding_react')) {
```
Comment (cableman) [inline modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: I'm not sure that I understand why code related to list migration is placed inside ding_adgangsplatformen. Why not introduce a `hook_ding_adgangsplatformen_post_login()` so this module can been kept clean?

File: modules/ding_react/ding_react.module [hunk 2]
```diff
@@ -356,24 +363,124 @@ function ding_react_admin_settings_form() {
...
+  }
+
+  $form['debug'] = [
+    '#type' => 'fieldset',
+    '#title' => t('Debugging'),
+  ];
+  $form['debug']['ding_react_debug'] = [
+    '#type' => 'checkbox',
+    '#title' => t('Enable debug logging'),
+    '#default_value' => variable_get('ding_react_debug'),
```
Comment (cableman) [inline modules/ding_react/ding_react.module]: `variable_get` vil default return `null` should it here have been `false`?

File: modules/ding_react/ding_react.module [hunk 3]
```diff
@@ -321,9 +322,10 @@ function ding_react_follow_searches_url() {
   return variable_get('ding_react_follow_searches_url', DING_REACT_FOLLOW_SEARCHES_STAGE_URL);
 }
 
-function ding_react_admin_settings_form() {
-  $form = [];
-
+/**
+ * Admin settings for the module.
+ */
+function ding_react_admin_settings_form($form, &$form_state) {
```
Comment (cableman) [inline modules/ding_react/ding_react.module]: Normally we have places these admin forms in their own files (`ding_react.admin.inc`) to keep the module file code more clean. Why the change of pattern here?

File: modules/ding_react/ding_react.module [hunk 4]
```diff
@@ -356,24 +363,124 @@ function ding_react_admin_settings_form() {
...
 function ding_react_element_validate_url($element, &$form_state, $form) {
   if (!empty($element['#value']) && !valid_url($element['#value'], true)) {
     form_error($element, t('Please enter a valid url.'));
   }
 }
 
+/**
+ * Admin settings form form submit handler.
+ *
+ * We do not use system_settings_form() here as our form is a mix of variables
```
Comment (cableman) [inline modules/ding_react/ding_react.module]: The `system_settings_form` also set a theme function and a message about the content being saved `The configuration options have been saved.` I don't seen these here?

File: modules/ding_react/ding_react.module [hunk 5]
```diff
@@ -356,24 +363,124 @@ function ding_react_admin_settings_form() {
...
+    '#title' => t('Debugging'),
+  ];
+  $form['debug']['ding_react_debug'] = [
+    '#type' => 'checkbox',
+    '#title' => t('Enable debug logging'),
+    '#default_value' => variable_get('ding_react_debug'),
+  ];
+
+  $form['actions']['#type'] = 'actions';
+  $form['actions']['submit'] = array('#type' => 'submit', '#value' => t('Save configuration'));
```
Comment (cableman) [inline modules/ding_react/ding_react.module]: Mixed array syntax used. Please use either `[ ]` or  `array()` !

File: modules/ding_react/ding_react.module [hunk 6]
```diff
@@ -384,65 +491,75 @@ function ding_react_user_migrate(stdClass $account) {
...
+    'headers' => [
+      'Authorization' => "Bearer ${token}",
+    ]
+  ];
 
-    // Add a prefix to show the user has been migrated. This way we do not
-    // throw away data yet but avoid migrating users multiple times.
-    $account->openlist_uid = $migrated_prefix . $legacy_uid;
-    user_save($account);
+  $material_list_url = variable_get('ding_react_material_list_url');
```
Comment (cableman) [inline modules/ding_react/ding_react.module]: Why are you not using the function `ding_react_material_list_url()` to get the URL, which also uses the fallback URL?

File: modules/ding_react/ding_react.module [hunk 7]
```diff
@@ -384,65 +491,75 @@ function ding_react_user_migrate(stdClass $account) {
...
+    );
+    return FALSE;
+  }
 
-    watchdog(
-      'ding_react',
-      'Migrated user data for material list og follow searches',
-      [],
-      WATCHDOG_NOTICE
+  $follow_searches_url = variable_get('ding_react_follow_searches_url');
```
Comment (cableman) [inline modules/ding_react/ding_react.module]: Why not use `ding_react_follow_searches_url()` to get the URL which has fallback URL?

File: modules/ding_react/ding_react.module (hunk 2 above)
Comment (kasperg) [reply modules/ding_react/ding_react.module]: In the current case it does not make a difference. The check interprets `NULL` and `FALSE` the same.

File: modules/ding_adgangsplatformen/ding_adgangsplatformen.module (hunk 1 above)
Comment (kasperg) [reply modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: I have added a new hook `ding_adgangsplatformen_login` and implemented this in `ding_react`. I opted for this name over `ding_adgangsplatformen_post_login` because it closer resembles the existing `hook_user_login`.

There is currently also a function called `ding_adgangsplatformen_login()` but that is not called when invoking the hook.
